packages/api/lambdas/sf-starter.js

```diff
@@ -27,33 +42,140 @@ function dispatch(message) {
+ * @returns {Promise}
+ * @throws {Error}
+ */
+async function incrementPrioritySemaphore(key, maximum) {
+  const semaphore = new Semaphore(
+    dynamodbDocClient(),
+    process.env.SemaphoresTable
+  );
+
+  try {
+    await semaphore.up(key, maximum);
+  } catch (err) {
+    if (err instanceof ResourcesLockedError) {
+      log.info(`The maximum number of executions for ${key} are already running. Could not start a new execution.`);
+    }
+    throw err;
+  }
+}
+
+/**
+ * Attempt to increment the priority semaphore and start a new execution.
+ *
+ * If `incrementPrioritySemaphore()` is unable to increment the priority semaphore,
+ * it throws an error and `dispatch()` is not called.
+ *
+ * @param {Object} queueMessage - SQS message
+ * @returns {Promise} - Promise returned by `dispatch()`
+ * @throws {Error}
+ */
+async function incrementAndDispatch(queueMessage) {
+  const message = get(queueMessage, 'Body', {});
+
+  if (!has(message, 'cumulus_meta.priorityKey')
+      || isNil(message.cumulus_meta.priorityKey)) {
+    throw new Error('Could not find priority key for message');
+  }
+
+  const priorityKey = get(message, 'cumulus_meta.priorityKey');
+  const priorityLevelInfo = get(message, `cumulus_meta.priorityLevels.${priorityKey}`, {});
+
+  const { maxExecutions } = priorityLevelInfo;
+  if (!maxExecutions) {
+    throw new Error(`Could not determine maximum executions for priority ${priorityKey}`);
+  }
+
+  await incrementPrioritySemaphore(priorityKey, maxExecutions);
+
+  return dispatch(message);
+}
+
+/**
+ * This is an SQS queue consumer.
  *
- * It reads messages from a given sqs queue based on the configuration provided
- * in the event object
+ * It reads messages from a given SQS queue based on the configuration provided
+ * in the event object.
  *
  * The default is to read 1 message from a given queueUrl and quit after 240
- * seconds
+ * seconds.
  *
  * @param {Object} event - lambda input message
  * @param {string} event.queueUrl - AWS SQS url
  * @param {string} event.messageLimit - number of messages to read from SQS for
  *   this execution (default 1)
  * @param {string} event.timeLimit - how many seconds the lambda function will
  *   remain active and query the queue (default 240 s)
- * @param {Object} _context - lambda context
- * @param {function} cb - lambda callback
- * @returns {undefined} - undefined
+ * @param {function} dispatchFn - the function to dispatch to process each message
+ * @param {number} visibilityTimeout - how many seconds messages received from
+ *   the queue will be invisible before they can be read again
+ * @returns {Promise} - A promise resolving to how many executions were started
+ * @throws {Error}
  */
-function handler(event, _context, cb) {
+async function handleEvent(event, dispatchFn, visibilityTimeout) {
   const messageLimit = event.messageLimit || 1;
   const timeLimit = event.timeLimit || 240;
```

Comment:
@mennovandiermen If `event.timeLimit` is set to `0`, this is going to default it to `240`.  Is that the expected behavior?  (0 evaluates to false in that conditional.)

Reply:
Probably not. We don't really do any special case handling of a `0` time limit. It polls once before checking the time left. I suppose if you wanted it to poll just once you could want to set it to 0.

Reply:
The `isNil` check used below is appropriate here.

Reply:
Do we want to support 0? Even though it would work as @mennovandiermen described, it seems like really 0 should mean that the function spends no time executing (which doesn't make a ton of sense) or that there is no time limit. Having a value of 0 where the function does spend some time executing seems confusing, at least to me.

Reply:
I don't support the idea of an endless/no-time limit execution of `message-consumer` because it may result in a case where Lambda terminates the execution the 5-minute mark after messages have been polled and scheduled but before they are deleted, resulting in duplicate executions. Other than that I don't have a preference.

Reply:
Yes, agreed. I was more commenting on the general meaning of 0 as a time parameter for execution

Reply:
Yes, I just wanted to be explicit on that point. A 0-second time limit execution executing at all does seem strange, and I'm not convinced that it's something we want to support.

Reply:
Whether or not we accept 0, I think the behavior should be explicit, not implicitly treating `0` as `false`.

```javascript
const timeLimit = get(event, 'timeLimit', 240);
```

I'll defer to you whether to allow zero.  If you decided to prohibit it, add a `if (timeLimit <= 0) then throw new Error(...);`